### PR TITLE
#12634. Fix for endless loop when closing a transfer with data left to write,…

### PR DESCRIPTION
**Pull request by mattw-mega**
… and the write fails

Also simplified the logic a little, with just a single try to write each piece of remaining reassembled file data.

### CODE REVIEW

File: include/mega/raid.h [hunk 1]
```diff
@@ -68,7 +68,7 @@ namespace mega {
         FilePiece* getAsyncOutputBufferPointer(unsigned connectionNum);
 
         // indicate that the buffer written by asyncIO (or synchronously) can now be discarded.
-        void bufferWriteCompleted(unsigned connectionNum);
+        void bufferWriteCompleted(unsigned connectionNum, bool failedAndAbandoned = false);
```
[the remark this answers is not in the document]
Comment (mattw-mega) [reply include/mega/raid.h]: clarified :-) also uncommented the 'override' keyword on the functions that had it commented before c++11

File: src/transferslot.cpp [hunk 2]
```diff
@@ -192,31 +197,32 @@ TransferSlot::~TransferSlot()
             }
         }
 
-        for (int j = 0; connections > 0; j++)
+        bool anydata = true;
```
[the remark this answers is not in the document]
Comment (mattw-mega) [reply src/transferslot.cpp]: updated :-)

File: src/raid.cpp [hunk 3]
```diff
@@ -308,15 +308,18 @@ RaidBufferManager::FilePiece* RaidBufferManager::getAsyncOutputBufferPointer(uns
 }
 
 
-void RaidBufferManager::bufferWriteCompleted(unsigned connectionNum)
+void RaidBufferManager::bufferWriteCompleted(unsigned connectionNum, bool failedAndAbandoned)
```
Comment (sergiohs84) [inline src/raid.cpp]:
```suggestion
void RaidBufferManager::bufferWriteCompleted(unsigned connectionNum, bool succeeded)
```

Comment (mattw-mega) [reply src/raid.cpp]: thanks, updated :-)

**Pull request merged**